build_tools/android/run_benchmarks.py

```diff
@@ -0,0 +1,460 @@
+                 capture_output=True).stdout.strip()
+
+
+def get_android_device_model(verbose=False):
+  """Returns the Android device model."""
+  model = execute(["adb", "shell", "getprop", "ro.product.model"],
+                  capture_output=True,
+                  verbose=verbose).stdout.strip()
+  model = re.sub(r"\W+", "-", model)
+  return model
+
+
+def get_android_cpu_abi(verbose=False):
+  """Returns the CPU ABI for the Android device."""
+  return execute(["adb", "shell", "getprop", "ro.product.cpu.abi"],
+                 capture_output=True,
+                 verbose=verbose).stdout.strip()
+
+
+def get_android_cpu_features(verbose=False):
+  """Returns the CPU features for the Android device."""
+  cpuinfo = execute(["adb", "shell", "cat", "/proc/cpuinfo"],
+                    capture_output=True,
+                    verbose=verbose).stdout.strip()
+  features = []
+  for line in cpuinfo.splitlines():
+    if line.startswith("Features"):
+      _, features = line.split(":")
+      return features.strip().split()
+  return features
+
+
+def get_android_gpu_name(verbose=False):
+  """Returns the GPU name for the Android device."""
+  vkjson = execute(["adb", "shell", "cmd", "gpu", "vkjson"],
+                   capture_output=True,
+                   verbose=verbose).stdout.strip()
+  vkjson = json.loads(vkjson)
+  name = vkjson["devices"][0]["properties"]["deviceName"]
+
+  # Perform some canonicalization:
+
+  # - Adreno GPUs have raw names like "Adreno (TM) 650".
+  name = name.replace("(TM)", "")
+
+  # Replace all consecutive non-word characters with a single hypen.
+  name = re.sub(r"\W+", "-", name)
+
+  return name
+
+
+class AndroidDeviceInfo(object):
+
+  def __init__(self, verbose=False):
+    self.model = get_android_device_model(verbose)
+    self.cpu_abi = get_android_cpu_abi(verbose)
+    self.gpu_name = get_android_gpu_name(verbose)
+    self.cpu_features = get_android_cpu_features(verbose)
+
+  def __str__(self):
+    features = ", ".join(self.cpu_features)
+    params = [
+        f"model='{self.model}'",
+        f"cpu_abi='{self.cpu_abi}'",
+        f"gpu_name='{self.gpu_name}'",
+        f"cpu_features=[{features}]",
+    ]
+    params = ", ".join(params)
+    return f"Android device <{params}>"
+
+  def get_arm_arch_revision(self):
+    """Returns the ARM architecture revision."""
+    if self.cpu_abi != "arm64-v8a":
+      raise ValueError("Unrecognized ARM CPU ABI; need to update the list")
+
+    # CPU features for ARMv8 revisions.
+    # From https://en.wikichip.org/wiki/arm/armv8#ARMv8_Extensions_and_Processor_Features
+    rev1_features = ["atomics", "asimdrdm"]
+    rev2_features = [
+        "fphp", "dcpop", "sha3", "sm3", "sm4", "asimddp", "sha512", "sve"
+    ]
+
+    rev = "ARMv8-A"
+    if any([f in self.cpu_features for f in rev1_features]):
+      rev = "ARMv8.1-A"
+    if any([f in self.cpu_features for f in rev2_features]):
+      rev = "ARMv8.2-A"
+    return rev
```

Comment:
It could be. But in Android land, pretty much all SoCs top at v8.2-A. See https://en.wikipedia.org/wiki/Comparison_of_ARMv8-A_cores (I don't think the Kryo 6XX having v8.4-A is correct there.) So we won't need to worry about it for now (and probably quite some time in the future). :)